### PR TITLE
Add listing details page

### LINKED ISSUE #16: Customer can view listing details

### PR DESCRIPTION
**Pull request by hzxcaryn**
Closes #16 
# Changes

- Added a listing details page that users are routed to when they click on a listing in the explore page or directly visit `/listing/:id`
- Wired up the back navigation in the listing details page when users click on the back button
  * If user came from the Explore page, they will go back in browser history
  * If user came from elsewhere (eg. from a shared link), clicking back will navigate them to the explore page
- `/listing/:id` first routes to `src/components/listing-details` before showing the contents of `src/components/customer/listing-details`.
  * I added this extra layer of abstraction because merchants would see slightly different content from customers, and we don't really want to bloat up the components by making it a shared page.
  * This shared page in `src/components/listing-details` should do a check for whether the visitor is a merchant or not, and redirect the user to the merchant page if the user is a merchant (Left as a TODO for now until merchant side page is up)

**Here is a screenshot of how it looks like:**
image

### CODE REVIEW

File: server/src/api/listings.ts [hunk 1]
```diff
@@ -29,7 +28,22 @@ listingRouter.get(
   '/',
   async (req: Request, res: Response, next: NextFunction) => {
     try {
-      const listings = await listingService.getAllListings();
+      const listing = await listingService.getAllListings();
```
Comment (ifhuang) [inline server/src/api/listings.ts:31]: listings

File: server/src/api/listings.ts [hunk 2]
```diff
@@ -29,7 +28,22 @@ listingRouter.get(
...
+);
+
+listingRouter.get(
+  '/:listingId',
+  async (req: Request, res: Response, next: NextFunction) => {
+    const {listingId: listingIdStr} = req.params;
+    const listingId = Number(listingIdStr);
+
+    try {
+      const listings = await listingService.getListing(listingId);
```
Comment (ifhuang) [inline server/src/api/listings.ts:46]: listing

File: web/src/components/common/DeadlineTag.tsx [hunk 3]
```diff
@@ -42,11 +48,11 @@ interface DeadlineTagProps {
  * is not over.
  * Displays number of days since deadline if listing has ended.
  */
-const DeadlineTag: React.FC<DeadlineTagProps> = ({deadline}) => {
+const DeadlineTag: React.FC<DeadlineTagProps> = ({deadline, largerFont}) => {
```
Comment (ifhuang) [inline web/src/components/common/DeadlineTag.tsx:51]: can you explain what is largerFont for in the comment? same for other files

File: web/src/Routes.tsx [hunk 4]
```diff
@@ -25,12 +25,18 @@ const CustomerExplorePage = lazy(() => import('components/customer/explore'));
 // Merchant Pages
 const MerchantLandingPage = lazy(() => import('components/merchant/landing'));
 const MerchantSignUpPage = lazy(() => import('components/merchant/sign-up'));
+// Shared Pages
```
Comment (ifhuang) [inline web/src/Routes.tsx:28]: can you add more comment about how is it shared?

File: web/src/Routes.tsx [hunk 5]
```diff
@@ -25,12 +25,18 @@ const CustomerExplorePage = lazy(() => import('components/customer/explore'));
...
+const ListingDetailsPage = lazy(() => import('components/listing-details'));
 // Design samples
 const DesignSamplesPage = lazy(() => import('components/design-samples'));
 
 const Routes: React.FC = () => (
   <Suspense fallback={<Loading />}>
     <Switch>
+      <CustomerCommitCountProvider>
```
Comment (ifhuang) [inline web/src/Routes.tsx:36]: we already have a <CustomerCommitCountProvider> below

File: web/src/api.ts [hunk 6]
```diff
@@ -33,3 +34,14 @@ export const getAllListings = async (): Promise<Listing[]> => {
   const res = await fetch(`${process.env.REACT_APP_SERVER_URL}/listings`);
   return res.json();
 };
+
+/**
+ * Fetches a particular listing with the specified customerId.
```
Comment (ifhuang) [inline web/src/api.ts:39]: listingId

File: server/src/api/listings.ts (hunk 1 above)
Comment (hzxcaryn) [reply server/src/api/listings.ts:31]: Oh I just realised this branch is outdated after I addressed my comments on the previous branch (which I merged)

File: web/src/components/common/DeadlineTag.tsx (hunk 3 above)
Comment (hzxcaryn) [reply web/src/components/common/DeadlineTag.tsx:51]: @ifhuang  Oh it just makes the component use a larger font for the text when the prop is passed in

File: web/src/Routes.tsx (hunk 4 above)
Comment (hzxcaryn) [reply web/src/Routes.tsx:28]: @ifhuang Going to a particular listing page(`https://gpay-group-buy.an.r.appspot.com/listing/:listingId`) is a common route for both merchants and customers. But our app will show the user to the correct component depending on whether they are a merchant or not.

File: web/src/Routes.tsx (hunk 5 above)
Comment (hzxcaryn) [reply web/src/Routes.tsx:36]: @ifhuang Where? It is the closing tag in line 39?

File: web/src/api.ts (hunk 6 above)
Comment (hzxcaryn) [reply web/src/api.ts:39]: oops! Good catch, thanks!

File: web/src/Routes.tsx (hunk 4 above)
Comment (ifhuang) [reply web/src/Routes.tsx:28]: yes. add some context to the comment in the codebase

File: web/src/Routes.tsx (hunk 5 above)
Comment (ifhuang) [reply web/src/Routes.tsx:36]: line 43

File: web/src/Routes.tsx (hunk 4 above)
Comment (hzxcaryn) [reply web/src/Routes.tsx:28]: @ifhuang So just something like this?
```suggestion
// Shared Pages (neither merchant nor customer specific)
```

File: web/src/Routes.tsx (hunk 5 above)
Comment (hzxcaryn) [reply web/src/Routes.tsx:36]: 😅 Oh just saw it. My bad for not realising that when dealing with all the merge conflicts earlier on! 😭

File: web/src/Routes.tsx (hunk 4 above)
Comment (ifhuang) [reply web/src/Routes.tsx:28]: how about:
Shared Pages. Different components may be displayed depending on whether user is a customer or merchant

Comment (hzxcaryn) [reply web/src/Routes.tsx:28]: Leaving the conversation we had on Chat here:

> Caryn Heng, 21 mins
I feel like it's not only pages like that that are considered Shared.
I only did the `// Customer Pages` and `// Merchant Pages` comments to organise how the imports look
and `// Shared Pages` is just pages that kind of belong to neither HAHA
Idk if the router (which only cares about routing url path to components) should be having comments such as "Different components may be displayed depending on whether user is a customer or merchant"
that feels like a comment that should be left for the ListingDetails shared component itself

> Caryn Heng, 12 mins
or does `// Common Pages` sound better? haha
I was thinking that other pages that are under the `// Shared Pages` comment can be pages whereby both of them see the exact same thing
1 simple example is the 404 page
And if I added the "Different components may be displayed depending on whether user is a customer or merchant" part the 404 page would not fit in there anymore

> Yifu Huang, 7 mins
ic
got your idea now. yeah then probably "Common Pages" sounds better


Will proceed to rename the comment to "Common Pages"!

File: web/src/components/listing-details/index.tsx [hunk 7]
```diff
@@ -0,0 +1,28 @@
+/**
```
Comment (xlai20) [inline web/src/components/listing-details/index.tsx:1]: I'm a bit confused by the folder organization. How is this "listing-details" different from the "customer/listing-details"? Why must it be in a separate folder here?

File: web/src/components/common/CommitProgress.tsx [hunk 8]
```diff
@@ -39,8 +40,10 @@ const CommitProgressContainer = styled.div`
         return 'column';
     }
   }};
+
   text-align: right;
-  font-size: 0.8em;
+  font-size: ${({largerFont}: CommitProgressContainerProps) =>
+    largerFont ? '1em' : '0.8em'};
```
Comment (xlai20) [inline web/src/components/common/CommitProgress.tsx:46]: In what situations are you using largerFont and normal font?

File: web/src/components/listing-details/index.tsx (hunk 7 above)
Comment (hzxcaryn) [reply web/src/components/listing-details/index.tsx:1]: @xlai20 Going to a particular listing page(https://gpay-group-buy.an.r.appspot.com/listing/:listingId) is a common route for both merchants and customers. The router will route the browser requesting for the listing page(https://gpay-group-buy.an.r.appspot.com/listing/:listingId) to the `components/listing-details` page.
However, we want merchants and customers to see different things when they are on https://gpay-group-buy.an.r.appspot.com/listing/:listingId. This is where we have `components/customer/listing-details` and `components/merchant/listing-details`. The `components/listing-details` page will be in charge of showing the visitor `components/customer/listing-details` or  `components/merchant/listing-details` depending on their identity (whether they are a customer or a merchant).

File: web/src/components/common/CommitProgress.tsx (hunk 8 above)
Comment (hzxcaryn) [reply web/src/components/common/CommitProgress.tsx:46]: @xlai20 Ah I will use largerFont in a full details page, and normal font in listings cards.

**Pull request merged**